src/server/qgsfcgiserverresponse.cpp

```diff
@@ -54,71 +56,130 @@ typedef struct QgsFCGXStreamData
+                                 QStringLiteral( "FCGIServer" ),                                                   //
+                                 Qgis::MessageLevel::Warning );
     }
   }
   else
   {
-    QgsMessageLog::logMessage( QStringLiteral( "FCGI_stdin is null! Socket monitoring disable." ), QStringLiteral( "FCGIServer" ), Qgis::MessageLevel::Warning );
+    QgsMessageLog::logMessage( QStringLiteral( "FCGI_stdout is null! Socket monitoring disabled." ), //
+                               QStringLiteral( "FCGIServer" ),                                       //
+                               Qgis::MessageLevel::Warning );
   }
 #endif
 }
 
+// Informs the thread to quit
+void QgsSocketMonitoringThread::stop()
+{
+  mShouldStop.store( true );
+  // Release the mutex so the try_lock in the thread will not wait anymore and
+  // the thread will end its loop as we have set 'mShouldStop' to true
+  mMutex.unlock();
+}
+
 void QgsSocketMonitoringThread::run()
 {
+  // Lock the thread mutex: every try_lock will take 333ms
+  mMutex.lock();
+
   if ( mIpcFd < 0 )
   {
     QgsMessageLog::logMessage( QStringLiteral( "Socket monitoring disabled: no socket fd!" ), QStringLiteral( "FCGIServer" ), Qgis::MessageLevel::Warning );
     return;
   }
 
 #if defined( Q_OS_UNIX ) && !defined( Q_OS_ANDROID )
+  const pid_t threadId = gettid();
+
+  mShouldStop.store( false );
   char c;
-  while ( !*mIsResponseFinished )
+
+  fd_set setOptions;
+  FD_ZERO( &setOptions );        // clear the set
+  FD_SET( mIpcFd, &setOptions ); // add our file descriptor to the set
+
+  struct timeval timeout;
+  timeout.tv_sec = 0;
+  timeout.tv_usec = 50000; // max 50ms of timeout for select
+
+  while ( !mShouldStop.load() )
   {
-    const ssize_t x = recv( mIpcFd, &c, 1, MSG_PEEK | MSG_DONTWAIT ); // see https://stackoverflow.com/a/12402596
-    if ( x < 0 )
-    {
-      // Ie. we are still connected but we have an 'error' as there is nothing to read
-      QgsDebugMsgLevel( QStringLiteral( "FCGIServer: remote socket still connected. errno: %1" ).arg( errno ), 5 );
-    }
-    else
+    int rv = select( mIpcFd + 1, &setOptions, NULL, NULL, &timeout ); // see https://stackoverflow.com/a/30395738
+    if ( rv == -1 )
     {
       // socket closed, nothing can be read
-      QgsDebugMsgLevel( QStringLiteral( "FCGIServer: remote socket has been closed! errno: %1" ).arg( errno ), 2 );
+      QgsMessageLog::logMessage( QStringLiteral( "FCGIServer %1: remote socket has been closed (select)! errno: %2" ) //
+                                   .arg( threadId )
+                                   .arg( errno ),
+                                 QStringLiteral( "FCGIServer" ), Qgis::MessageLevel::Warning );
       mFeedback->cancel();
       break;
     }
+    else
+    {
+      const ssize_t x = recv( mIpcFd, &c, 1, MSG_PEEK | MSG_DONTWAIT ); // see https://stackoverflow.com/a/12402596
```

Comment:
could you add a comment to explain this case and the difference with the one above

Reply:
done!